src/services/codefixes/fixClassDoesntImplementInheritedAbstractMember.ts

```diff
@@ -41,7 +41,9 @@ namespace ts.codefix {
         // so duplicates cannot occur.
         const abstractAndNonPrivateExtendsSymbols = checker.getPropertiesOfType(instantiatedExtendsType).filter(symbolPointsToNonPrivateAndAbstractMember);
 
-        createMissingMemberNodes(classDeclaration, abstractAndNonPrivateExtendsSymbols, context, preferences, member => changeTracker.insertNodeAtClassStart(sourceFile, classDeclaration, member));
+        const importAdder = createImportAdder(sourceFile, context.program, preferences, context.host);

```

Comment:
This pattern feels like dynamic scope (or, a specialisation of the state monad) that's not enforced by the compiler. I'm not a fan, although I don't know an easy way to improve it. Is it really necessary to be able to write the fixes out at the end? (the code in inferFromUsage is an uglier example of this)

Reply:
> Is it really necessary to be able to write the fixes out at the end?

Yes—the whole point of this PR is the acknowledgment that to auto-import more than one name at a time, you have to consider them together in order to combine and deduplicate them properly. I think part of the problem is that `changeTracker` is _also_ a state container that gets passed around getting state incrementally tacked onto it, so we’re already operating in a mode where deep call stacks mutate shared state and return nothing. And what’s more, that existing state lacks any semantic meaning that would let us derive the necessary imports changes from it (i.e., ChangeTracker works only in strings and nodes, not in symbols or types, which are needed for auto imports). Hence, another mutable state container to complement `ChangeTracker`.